nomos-services/chain/chain-leader/src/leadership.rs

```diff
@@ -124,4 +101,139 @@ impl Leader {
 
         None
     }
+
+    fn private_inputs_for_winning_utxo_and_slot(
+        &self,
+        utxo: &Utxo,
+        epoch: Epoch,
+        public_inputs: LeaderPublic,
+    ) -> LeaderPrivate {
+        // TODO: Get the actual witness paths and leader key
+        let aged_path = Vec::new(); // Placeholder for aged path
+        let latest_path = Vec::new();
+        let slot_secret = *self.sk.as_fr();
+        let starting_slot = self.config.epoch_config.starting_slot(&epoch).into();
+        let leader_pk = ed25519_dalek::VerifyingKey::from_bytes(&[0; 32]).unwrap(); // TODO: get actual leader public key
+
+        LeaderPrivate::new(
+            public_inputs,
+            *utxo,
+            &aged_path,
+            &latest_path,
+            slot_secret,
+            starting_slot,
+            &leader_pk,
+        )
+    }
+
+    fn slot_secret_key(&self, _slot: Slot) -> SecretKey {
+        self.sk.clone()
+    }
+}
+
+fn public_inputs_for_utxo(
+    utxo: &Utxo,
+    epoch_state: &EpochState,
+    slot: Slot,
+    latest_tree: &UtxoTree,
+) -> Option<LeaderPublic> {
+    let () = epoch_state.utxos.witness(&utxo.id())?;
+    let () = latest_tree.witness(&utxo.id())?;
```

Comment:
These would be used in the `private_inputs_..` above, they would fill in the `aged_path` and `latest_path` placeholder values.

Reply:
Ah, its already here. So then better to fill. Thanks David.

Reply:
Yes but `witness` returns a `()` while the path is expected to be a vector. I did not want to change too much logic in this PR, so I'd feel more comfortable to tackle this change specifically in a separate one, what do you guys think?